resourceIndexer/coreStagingS3.js

```diff
@@ -0,0 +1,292 @@
+import {logger} from './logger.js';
+import { SOURCE, VALID_AUSPICE_PATTERNS, INVALID_AUSPICE_PATTERNS,
+  DATESTAMP_REGEX, MAIN_DATASET_JSON } from './constants.js';
+
+/**
+ * The inventory of buckets (especially the core bucket) is in some ways a
+ * historical record of work over the years, but this isn't really what we want
+ * to display to users. As some examples:
+ *    - Files which don't match a resource to list should be excluded
+ *    - Datestampted files (i.e. _YYYY-MM-DD in the filename) are excluded
+ *      (we use S3 versioning instead)
+ * 
+ * If the s3 object is to be excluded we return false here.
+ * 
+ * In the case where the object represents a (part of) a resource we want to
+ * expose, then we categorise it here by adding the following properties:
+ *   - source (STAGING or CORE)
+ *   - resourceType (dataset, narrative or intermediate)
+ *   - id (the ID by which objects will be grouped together.
+ *        For datasets this is the nextstrain.org URL path, without any temporal signifier)
+ *   - subresourceType (currently only for resourceType=dataset)
+ */
+function categoriseCoreObjects(item, staging) {
+  const key = item.key
+  item.source = staging ? SOURCE.STAGING : SOURCE.CORE;
+  item.baseUrl = `https://nextstrain-${staging?'staging':'data'}.s3.amazonaws.com/${key}`
```

Comment:
Ok. I don't have a problem with this change, so I've adopted it to help get this PR merged, but I don't see how it's really an improvement. Is this to facilitate changing the bucket name of core and/or staging? Could you say more here?

Reply:
It's about keeping boundaries between layers clear and tight rather than tangled. Tangled boundaries make code harder to maintain and modify over time.

An S3 object is intrinsically in a bucket, and that bucket is given right there in the inventory record. An object's bucket is not intrinsically tied to some boolean passed to a function. It may be inferable due to the current logic of those function calls and the boolean passed, but that's all subject to change. The intrinsic link between an object and bucket is not.

> Is this to facilitate changing the bucket name of core and/or staging? 

That's not the main motivation for my suggestion, but yes, it would facilitate that. For example, consider if in automated tests we start backing the `CoreSource` with nextstrain-testing-data instead of nextstrain-data so we don't have to worry about destructive actions in tests.